portal/account/account.lib.php

```diff
@@ -152,42 +151,52 @@ function doCredentials($pid)
     global $srcdir;
     require_once("$srcdir/authentication/common_operations.php");
 
-    $newpd = sqlQuery("SELECT * FROM `patient_data` WHERE `pid`=?", array(
+    $newpd = sqlQuery("SELECT id,fname,mname,lname,email,email_direct FROM `patient_data` WHERE `pid`=?", array(
         $pid
     ));
 
+    $uname = $newpd['fname'] . $newpd['id'];
+    // Token expiry 1 hour
+    $expiry = new DateTime('NOW');
+    $expiry->add(new DateInterval('PT01H'));
+
     $clear_pass = generatePassword();
 
-    $uname = $newpd['fname'] . $newpd['id'];
+    $token = openssl_random_pseudo_bytes(32);
```

Comment:
We have a more secure ways to create random bytes or tokens here. First does random bytes (this is truly random and secure; above openssl_random_pseudo_bytes() function is no longer considered random or secure). Second provides flexible characters used. Third only uses alphanumeric characters:
https://github.com/openemr/openemr/blob/master/src/Common/Utils/RandomGenUtils.php#L18
https://github.com/openemr/openemr/blob/master/src/Common/Utils/RandomGenUtils.php#L34
https://github.com/openemr/openemr/blob/master/src/Common/Utils/RandomGenUtils.php#L53
(check out codebase for examples of use)

Reply:
lol, I knew these were there but I couldn't find them!!!! Tired eyes, thanks, i'll change.  Can I have today to get this back to 502 and maybe a couple more bugs in portal or are you working patch now?

Reply:
@bradymiller Oh also, very nice work on the portal session fixups you did. They nicely fixed some issues i've seen in past with the several pre login scripts (register, reset etc). So thanks.